### ISSUE TITLE
Model saving much slower with 2.0?

### ISSUE THREAD

**Issue by pmichel31415** [clab/dynet#648]
Saving a medium sized seq2seq model takes between 2x and 3x longer as before, is anyone else experiencing something similar?

**Comment by neubig** [comment]
Hmm, maybe @xunzhang can take a look at this?

**Comment by xunzhang** [comment]
Hi @pmichel31415 , I did a performance regression test just now. But in my Linux environment, the save function in v2.0(12s, 266MB model file after saving) is a little faster than in dynet1.1(14s, 525MB model file after saving). For load, since the model file in v2.0 is much smaller, the loading time much shorter(8s vs 14.5s). I tested both cpp and python interface, there is no obvious difference. I also tested saving with nested subcollection, the result is similar.
A possible possibility is that I generated parameters and lookup parameters randomly while you run a real model data. So could you give the dataset and how to reproduce it so that I can tune it, thanks.

**Comment by pmichel31415** [comment]
Hi @xunzhang , thanks for looking into this.

I've noticed that for some other runs the save time was much faster (for the same model/data) sot it might be a problem if read/write speed on my machine?

In any case, do you think it would be hard/a lot of work to implement binary saving/loading with the current interface? I would be willing to code that.

**Comment by neubig** [comment]
I don't think implementing a binary serializer would be too hard, and it would certainly be useful.

**Comment by xunzhang** [comment]
@pmichel31415, I think it shouldn't be too difficult. It is useful if the performance is much faster using binary serializer.

**Comment by pmichel31415** [comment]
Hmm so I inspected the model files manually and I saw a lot of zeros. 

It appears that the `TextFileSaver` object is saving the gradients as well as the values.

I can see the advantage of saving gradients on some cases but I think most of the times people save the parameters after updating them (so the gradients are 0).

Maybe there should be a a flag to enable gradient saving, or some way to indicate that the gradients are 0 and that they don't need to be saved.

That would save a lot of space.

**Comment by yoavg** [comment]
@pmichel31415 I think asking the user to save gradients or not is kinda inelegant. I think an easier solution would be to always save the gradients, but if they are all zero just indicate it with a predefined symbol (say "ZERO_GRAD") instead of a long list of zeros. Then loading needs to be aware of that, and that's it. We'll get both the smaller model and speedier load/save times for this case, without any effort from the user.

Re binary format: should be trivial to do, but need to make sure it is compatible across machines.
I also suggest using a hybrid format: keep most of the format as it is now, but just replace the weights themselves with a run of binary values.

**Comment by pmichel31415** [comment]
Yes I agree this is the best solution, I just did a pull request to this effect #663
